### PR TITLE
Add expected start and end date to availability

### PR DESCRIPTION
**Pull request by efronbs**
issue: https://github.com/yahoo/fili/issues/822

PLEASE READ:
This PR changes the algorithm for how `PartitionAvailability` reports its available intervals. For all existing cases the output should be the same. The tests are the most important part of this PR. Please spend most of your time reviewing the new tests. The two new tests in `PartitonAvailabilitySpec` are the most important tests.

The first picture is the tests for finding missing intervals, and the second picture is for testing `getAvailabilities(constraint)` with start and end dates. The tests are in the same order as in the pictures.

Visualization of first block of tests in PartitionAvailabilitySpec, in order

Visualization of second block of tests in PartitionAvailabilitySpec, in order

### CODE REVIEW

File: fili-core/src/test/groovy/com/yahoo/bard/webservice/table/availability/BaseCompositeAvailabilitySpec.groovy [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+
+        where:
+        testExpectedStart_1 |   testExpectedStart_2 |   testExpectedStart_3 |   desc
+        earliestStart       |   middleStart         |   null                |   "one"
+        earliestStart       |   null                |   null                |   "two"
+        null                |   null                |   null                |   "all"
+    }
+
+    @Unroll
+    def "If #desc availabilities have no end date, the composite has no end date"() {
```
Comment (michael-mclawhorn) [inline fili-core/src/test/groovy/com/yahoo/bard/webservice/table/availability/BaseCompositeAvailabilitySpec.groovy]: This test doesn't DO anything with it's parameters.

File: fili-core/src/test/groovy/com/yahoo/bard/webservice/table/availability/BaseCompositeAvailabilitySpec.groovy [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+        testExpectedEnd_1   |   testExpectedEnd_2   |   testExpectedEnd_3   |   desc
+        earliestEnd         |   middleEnd           |   null                |   "one"
+        earliestEnd         |   null                |   null                |   "two"
+        null                |   null                |   null                |   "all"
+    }
+
+    @Unroll
+    def "earliest start is at position #pos in the availabilities stream"() {
+        given:
+        expectedStart_1 = Optional.ofNullable((DateTime) testExpectedStart_1)
```
Comment (michael-mclawhorn) [inline fili-core/src/test/groovy/com/yahoo/bard/webservice/table/availability/BaseCompositeAvailabilitySpec.groovy]: This test also seems to not do anything with the arguments.

File: fili-core/src/test/groovy/com/yahoo/bard/webservice/table/availability/BaseCompositeAvailabilitySpec.groovy (hunk 1 above)
Comment (efronbs) [reply fili-core/src/test/groovy/com/yahoo/bard/webservice/table/availability/BaseCompositeAvailabilitySpec.groovy]: A lot of the logic is happening in the setup function. `compositeAvailability` is being constructed out of 3 different Availability mocks. The parameters are setting the availability of those 3 mocks, and I am checking the behavior of the composite availability that wraps them.

File: fili-core/src/test/groovy/com/yahoo/bard/webservice/table/availability/BaseCompositeAvailabilitySpec.groovy (hunk 2 above)
Comment (efronbs) [reply fili-core/src/test/groovy/com/yahoo/bard/webservice/table/availability/BaseCompositeAvailabilitySpec.groovy]: same as above

**Pull request merged**